### PR TITLE
Added Topological sort

### PR DESCRIPTION
**Pull request by liach**
The topological sort thing has two addon features: 
1. Reporting all the cycles on failure (Using [tarjan's strongly connected component algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm))
2. Sorting based on natural order/comparator in addition to the topological order. May be useful for events or plugins that both define a dependency order and an arbitrary integer priority, etc.

Moreover, I added an annotation for `Optionals` class since intellij idea was crying for varargs.

### CODE REVIEW

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * A collection of a few graph algorithms.
+ */
+public final class GraphAlgorithms {
+
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: remove new line

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+  /**
+   * Sorts a directed acyclic graph into a list.
+   * The particular order of elements without prerequisites is not guaranteed.
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: new line, wrap in `<p></p>`

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+   *
+   * @param graph the graph to be sorted
+   * @param <T>   the node type
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: too many spaces

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+  @NonNull
+  public static <T> List<T> topologicalSort(final @NonNull Graph<T> graph) {
+    return topologicalSort(graph, randomSortType());
+  }
+
+  /**
+   * Sorts a directed acyclic graph into a list.
+   * The particular order of elements without prerequisites is determined by the comparator.
+   *
+   * @param graph      the graph to be sorted
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: too many spaces

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  public static <T> List<T> orderedTopologicalSort(final @NonNull Graph<T> graph, final @NonNull Comparator<? super T> comparator) {
+    return topologicalSort(graph, new ComparatorSortType<>(comparator));
+  }
+
+  /**
+   * Sorts a directed acyclic graph into a list.
+   * The particular order of elements without prerequisites is determined by the natural order.
+   *
+   * @param graph the graph to be sorted
+   * @param <T>   the node type, implementing {@link Comparable}
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: too many spaces

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+    if(!requiredCounts.isEmpty())
+      throw new IllegalArgumentException("Graph (" + graph + ") has cycle(s): " + new StronglyConnectedComponentAnalyzer<>(graph).analyze());
+
+    return results;
+  }
+
+  @VisibleForTesting
+  static final class StronglyConnectedComponentAnalyzer<T> {
+    @NonNull
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: annotations on same line as field

File: src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    private final Deque<T> stack;
+    @NonNull
+    private final StringBuilder reporter;
+    @NonNegative
+    private int currentId;
+
+    @VisibleForTesting
+    StronglyConnectedComponentAnalyzer(@NonNull final Graph<T> graph) {
+      this.graph = graph;
+      this.ids = new HashMap<>();
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/GraphAlgorithms.java]: can be on the field itself

File: src/test/java/net/kyori/lunar/graph/GraphAlgorithmsTest.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import com.google.common.graph.MutableGraph;
+import org.junit.Test;
+
+import java.util.Comparator;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class GraphAlgorithmsTest {
+
```
Comment (kashike) [inline src/test/java/net/kyori/lunar/graph/GraphAlgorithmsTest.java]: remove new line

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+   * Sorts a directed acyclic graph into a list.
+   * <p>The particular order of elements without prerequisites is not guaranteed.</p>
+   *
+   * @param graph the graph to be sorted
+   * @param <T> the node type
+   * @return the sorted list
+   * @throws IllegalArgumentException if the graph is not directed, allows self loops, or has cycles
+   */
+  @NonNull
+  public static <T> List<T> topologicalSort(final @NonNull Graph<T> graph) {
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: `@NonNull final`

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+/**
+ * A collection of a few graph algorithms.
+ */
+public final class MoreGraphs {
+  /**
+   * Sorts a directed acyclic graph into a list.
+   * <p>The particular order of elements without prerequisites is not guaranteed.</p>
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: New line before this line, please.

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+   * @throws IllegalArgumentException if the graph is not directed, allows self loops, or has cycles
+   */
+  @NonNull
+  public static <T> List<T> topologicalSort(final @NonNull Graph<T> graph) {
+    return topologicalSort(graph, SortType.random());
+  }
+
+  /**
+   * Sorts a directed acyclic graph into a list.
+   * <p>The particular order of elements without prerequisites is determined by the comparator.</p>
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: New line before this line, please.

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+   * @throws IllegalArgumentException if the graph is not directed, allows self loops, or has cycles
+   */
+  @NonNull
+  public static <T extends Comparable<? super T>> List<T> orderedTopologicalSort(@NonNull final Graph<T> graph) {
+    return topologicalSort(graph, SortType.comparable());
+  }
+
+  @NonNull
+  private static <T> List<T> topologicalSort(@NonNull final Graph<T> graph, @NonNull final SortType<T> sortType) {
+    checkArgument(graph.isDirected(), "The graph must be directed!");
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: all lowercase, remove `!`

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    checkArgument(!graph.allowsSelfLoops(), "The graph cannot allow self loops!");
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: all lowercase, remove `!`

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+          processing.add(successor);
+          requiredCounts.remove(successor);
+        } else {
+          requiredCounts.put(successor, newCount);
+        }
+      }
+      results.add(now);
+    }
+
+    if(!requiredCounts.isEmpty())
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: braces

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+      throw new IllegalArgumentException("Graph (" + graph + ") has cycle(s): " + new StronglyConnectedComponentAnalyzer<>(graph).printCycles());
+
+    return results;
+  }
+
+  private MoreGraphs() {
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: move to top of class

File: src/main/java/net/kyori/lunar/graph/SortType.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  @NonNull
+  Queue<T> makeQueue();
+
+  @SuppressWarnings("unchecked")
+  @NonNull
+  static <T> SortType<T> random() {
+    return (SortType<T>) Random.INSTANCE;
+  }
+
+  final class Random implements SortType<Object> {
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/SortType.java]: Move each of these to their own class - `RandomSortType`, `ComparableSortType`, etc.

File: src/main/java/net/kyori/lunar/graph/ComparatorSortType.java [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+import net.kyori.blizzard.NonNull;
+
+import java.util.Comparator;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+final class ComparatorSortType<T> implements SortType<T> {
+  private final Comparator<? super T> comparator;
+
+  ComparatorSortType(Comparator<? super T> comparator) {
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/ComparatorSortType.java]: `final`

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+  @NonNull
+  public static <T extends Comparable<? super T>> List<T> orderedTopologicalSort(@NonNull final Graph<T> graph) {
+    return topologicalSort(graph, SortType.comparable());
+  }
+
+  /**
+   * Actual content of the topological sort. This is a breadth-first search based approach.
+   */
+  @NonNull
+  private static <T> List<T> topologicalSort(@NonNull final Graph<T> graph, @NonNull final SortType<T> sortType) {
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: `sortType` -> `type`

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+   * @return the sorted list
+   * @throws IllegalArgumentException if the graph is not directed, allows self loops, or has cycles
+   */
+  @NonNull
+  public static <T extends Comparable<? super T>> List<T> orderedTopologicalSort(@NonNull final Graph<T> graph) {
+    return topologicalSort(graph, SortType.comparable());
+  }
+
+  /**
+   * Actual content of the topological sort. This is a breadth-first search based approach.
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: Missing `@param` for `graph` and `type`.

File: src/main/java/net/kyori/lunar/graph/SortType.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.Queue;
+
+/**
+ * The type of sorts performed by the topological sorter.
+ * <p>This is not an enum because of generics.</p>
+ */
+@FunctionalInterface
+interface SortType<T> {
+  @NonNull
+  Queue<T> makeQueue();
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/SortType.java]: `createQueue`

File: src/main/java/net/kyori/lunar/graph/SortType.java [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+  SortType<?> RANDOM = ArrayDeque::new;
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/SortType.java]: fields to the top of the class

File: src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+  StronglyConnectedComponentAnalyzer(@NonNull final Graph<T> graph) {
+    this.graph = graph;
+    this.low = new int[graph.nodes().size()];
+  }
+
+  @VisibleForTesting
+  void analyze() {
+    for(final T node : this.graph.nodes()) {
+      if(!this.ids.containsKey(node)) {
+        dfs(node);
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java]: `this.`

File: src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+  }
+
+  @NonNull
+  @VisibleForTesting
+  Collection<Collection<T>> getComponents() {
+    return this.components;
+  }
+
+  @NonNull
+  String printCycles() {
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java]: `renderCycles` or similar

File: src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    final int self = this.currentId;
+    this.low[self] = self;
+    this.stack.push(now);
+    this.currentId++;
+
+    for(final T node : this.graph.successors(now)) {
+      if(this.ids.containsKey(node)) {
+        this.low[self] = Math.min(this.ids.get(node), this.low[self]);
+      } else {
+        dfs(node);
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java]: `this.`

File: src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java (hunk 22 above)
Comment (liach) [reply src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java]: I thought only fields need qualifiers.

File: src/main/java/net/kyori/lunar/graph/MoreGraphs.java (hunk 19 above)
Comment (liach) [reply src/main/java/net/kyori/lunar/graph/MoreGraphs.java]: Fixed.

File: src/main/java/net/kyori/lunar/graph/SortType.java [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+  @NonNull
+  Queue<T> createQueue();
+
+  /**
+   * Gets the type-safe instance of the random sort type.
+   *
+   * @param <T> the node type
+   * @return the random sort type
+   */
+  @SuppressWarnings("unchecked")
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/SortType.java]: move onto line after `@NonNull`

File: src/main/java/net/kyori/lunar/graph/SortType.java [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+    return (SortType<T>) RANDOM;
+  }
+
+  /**
+   * Gets the type-safe instance of the comparable sort type.
+   *
+   * @param <T> the node type
+   * @return the comparable sort type
+   */
+  @SuppressWarnings("unchecked")
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/SortType.java]: move onto line after `@NonNull`

File: src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+ * <p>This is based on Tarjan's Strongly Connected Component algorithm.</p>
+ *
+ * @param <T> the graph's node type
+ */
+@VisibleForTesting
+final class StronglyConnectedComponentAnalyzer<T> {
+  @NonNull private final Map<T, Integer> ids = new HashMap<>();
+  @NonNull private final Deque<T> stack = new ArrayDeque<>();
+  @NonNull private final Collection<Collection<T>> components = new ArrayList<>();
+  @NonNegative private int currentId = 0;
```
Comment (kashike) [inline src/main/java/net/kyori/lunar/graph/StronglyConnectedComponentAnalyzer.java]: move below `low` field

**Comment by liach** [comment]
Oh dear... The merge is fast and furious.

**Pull request merged**